### DISCUSSION TITLE
Bulk Loading Tools + SSH Tunnel

### DISCUSSION THREAD

**Discussion by isaacrlee** [slingdata-io/sling-cli#279] [Q&A] [answered]
If `use_bulk` is set to `true`, a bulk loading tool such as `bcp` is installed, and `ssh_tunnel` and `ssh_private_key` are set, would the connection to the SSH tunnel be used by bcp?

**Comment by flarco** [comment]
Ah, interesting premise. No, `bcp` is only checked and used locally.
See https://github.com/slingdata-io/sling-cli/blob/main/core/dbio/database/database_sqlserver.go#L197.
Actually, thinking of it, even if ssh is set, if bcp is detected locally, sling will forget about the ssh and attempt to load from the local machine. Not ideal.
This logic needs to be implemented. Opened https://github.com/slingdata-io/sling-cli/issues/280 for this.

**Reply by flarco** [reply]
@isaacrlee to be clear, `bcp` would need to be installed on the ssh machine. As far as I understand, bcp does not connect to ssh to import into SQL server.

**Reply by isaacrlee** [reply]
Thanks for the reply. I might be misunderstanding Sling/`bcp`/SSH tunnels, but would it not be possible for `bcp` to use an updated connection URL with local host and port after the SSH tunnel is created?

I see the code for the updated connection URL used by later non-bulk inserts [here](https://github.com/slingdata-io/sling-cli/blob/39bf63a26cd95a98e33424f342286ce0719bf487/core/dbio/database/database.go#L581), but [here](https://github.com/slingdata-io/sling-cli/blob/39bf63a26cd95a98e33424f342286ce0719bf487/core/dbio/database/database_sqlserver.go#L380) the original connection URL will always be used for BCP imports.

**Reply by flarco** [reply]
Actually, that's a good one. Instead of having bcp on the ssh machine, bcp should simply point to the forwarded port. Good.

**Reply by flarco** [reply, marked as answer]
Added to branch `v1.2.7`

**Reply by flarco** [reply]
https://github.com/slingdata-io/sling-cli/pull/274/commits/ecd9bdb33378d8a2aecb75864e56675edb0f9a00

**Reply by isaacrlee** [reply]
Awesome, thank you!
